projects/resources-and-entities.md

```diff
@@ -0,0 +1,122 @@
+
+>For example, a process producing telemetry that is running in a container on Kubernetes has a Pod name, it is in a namespace and possibly is part of a Deployment which also has a name. All three of these attributes can be included in the Resource.
+
+The problem with such usage is that by looking at the Resource attributes it is impossible to tell which of the represented entities is *the entity*, i.e. the entity which produced the telemetry.
+
+### Problem 2: Lack of Precise Identity
+
+The Resource is one set of attributes, which contains all attributes of all entities that the Resource represents. It is impossible to tell which of these attributes identify the entity (or entities) and which are non-identifying, i.e. purely descriptive.
+
+This lack of precise identity makes it difficult or impossible to identify the same entities reported in different Resources. 
+
+### Problem 3: Lack of Mutable Attributes
+
+Resource is defined to be immutable in the OpenTelemetry SDK. This does not align well with the fact that non-identifying attributes of entities may change over time. For example OpenTelemetry Collector collects data about Pods and adds Pod labels as Resource attributes. Pod labels are mutable in Kubernetes and can change over time, while the Pod's identity remains immutable. Here is another example where [mutable Service attributes](https://github.com/open-telemetry/opentelemetry-specification/issues/3401#issuecomment-1511770735) are desirable.
+
+With the current definition of the Resource we are forced to either leave out any attributes that may ever change over time or violate the spec definition.
+
+Additionally, OpenTelemetry currently lacks the ability to provide resource attributes that require some kind of delayed lookup that may fail (see [this issue](https://github.com/open-telemetry/opentelemetry-specification/issues/1298)). This required, e.g. passing environment variables for k8s container name and various downward-api values for an OpenTelemetry SDK to appropriately report this resource.
+
+In reality, OpenTelemetry SDKs can also easily violate the definition as soon as we consider mutability from a recipient's perspective. SDKs only guarantee immutability during a single process session. As soon as the process is restarted and the SDK is newly initialized, there is no guarantee that the Resource will have the same set of attributes (e.g. because `process.id` can be one of the Resource attributes).
+It is clear that the strictly "immutable" definition of the Resource is not sufficient for what we are trying to model.
+
+### Problem 4: Metric Cardinality Problem
+
+Today every attribute in an OpenTelemetry Resource, according to the metric data model, is used to determine the identity of a metric.  Given known issues in metric time-series database implementations around cardinality, this can cause major issues if Resources are allowed to leverage high cardinality attributes. 
```

Comment:
Is it really true that every resource attribute is supposed to determine its identity? In OTLP, the `service.name`, `service.namespace`, and `service.instance.id` triplet is [required to be unique](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#resource-attributes-1), and is used to identify the resource when constructing `target_info` in Prometheus.

Reply:
>Is it really true that every resource attribute is supposed to determine the resource's identity?

Yes, metric objects are identified by several pieces of data, including Resource attributes, [see here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#opentelemetry-protocol-data-model).

Quote from the spec:

>Metric streams are grouped into individual Metric objects, identified by:
> - The originating Resource attributes
> - ...

Reply:
>In OTLP, the service.name, service.namespace, and service.instance.id triplet is [required to be unique](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#resource-attributes-1), and is used to identify the resource when constructing target_info in Prometheus (becoming the identifying job and instance labels).

That is not required in OTLP, the Prometheus compatibility doc is wrong. The uniqueness of the triplet is merely a semantic convention for Services. Not every Resource is a Service in OpenTelemetry.

Reply:
Filed this to fix the Prometheus compatibility doc: https://github.com/open-telemetry/opentelemetry-specification/issues/3944

Reply:
OK, thanks for clarifying that!